File: JavaFx/JavaFx/src/main/java/com/minkyu/yourdailyword/javafx/models/IQuotesManager.java
package com.minkyu.yourdailyword.javafx.models;

import com.minkyu.yourdailyword.common.calendar.ICalendarModel;
import com.minkyu.yourdailyword.common.protobased.*;
import com.minkyu.yourdailyword.common.protos.Quotes;
import org.jetbrains.annotations.Nullable;

import java.util.Locale;
import java.util.function.Consumer;

public interface IQuotesManager {
    int getNewUid();

    void addQuote(QuoteModel quoteModel);

    void setFromProto(Quotes newProto);

    Quotes getProto();

    int getNumOfQuotes();

    @Nullable
    String getQuote(ICalendarModel calendarModel);

    CalendarTypeModel getCalendarType();

    @Nullable
    QuoteModel getQuoteModelByIndex(int index);

    void getQuoteModelByIndexAndDoIfNotNull(int index, Consumer<QuoteModel> action);

    @Nullable
    QuoteModel getQuoteModelByUid(int uid);

    void getQuoteModelByUidAndDoIfNotNull(int uid, Consumer<QuoteModel> action);

    void deleteQuoteModelByUid(int uid);

    @Nullable
    String getQuoteBasedOnDayOfMonth(int month, int dayOfMonth, Locale locale);

    @Nullable
    GregorianCalendarOptionsModel getGregorianCalendarOptions();

    @Nullable
    LunarCalendarOptionsModel getLunarCalendarOptionsModel();

    @Nullable
    HebrewCalendarOptionsModel getHebrewCalendarOptions();

    Runnable addAfterQuotesUpdateRunnable(Runnable runnable);

    void saveCurrentModelToFile();
}

File: Android/app/src/main/java/com/minkyu/yourdailyword/android/models/IQuotesManager.kt
package com.minkyu.yourdailyword.android.models

import com.minkyu.yourdailyword.common.protobased.CalendarTypeModel
import com.minkyu.yourdailyword.common.protobased.QuoteModel
import com.minkyu.yourdailyword.common.protobased.QuotesModel
import com.minkyu.yourdailyword.common.protos.Quotes
import kotlinx.coroutines.flow.MutableStateFlow

@Suppress("TooManyFunctions")
interface IQuotesManager { val modelFlow: MutableStateFlow<QuotesModel?>
	suspend fun setFromProto(quotesProto: Quotes)
	fun getQuotesCalendarType(): CalendarTypeModel
	fun getQuote(quoteId: Int): QuoteModel?
	fun getProto(): Quotes?
	suspend fun deleteQuote(quoteId: Int)
	suspend fun updateQuote(quoteModel: QuoteModel)
	suspend fun saveToDefaultFile()
	suspend fun addQuote(quoteModel: QuoteModel)
	fun getNewUid(): Int
}
